# Homework Basic Python
** Author : Ferry Wyman

In [100]:
import pandas as pd
import numpy as np

In [103]:
gps = pd.read_csv ('D:\Chapter 5\DSA\googleplaystore.csv')

In [104]:
gps.head (100)

App           Category  \
0      Photo Editor & Candy Camera & Grid & ScrapBook     ART_AND_DESIGN   
1                                 Coloring book moana     ART_AND_DESIGN   
2   U Launcher Lite – FREE Live Cool Themes, Hide ...     ART_AND_DESIGN   
3                               Sketch - Draw & Paint     ART_AND_DESIGN   
4               Pixel Draw - Number Art Coloring Book     ART_AND_DESIGN   
5                          Paper flowers instructions     ART_AND_DESIGN   
6             Smoke Effect Photo Maker - Smoke Editor     ART_AND_DESIGN   
7                                    Infinite Painter     ART_AND_DESIGN   
8                                Garden Coloring Book     ART_AND_DESIGN   
9                       Kids Paint Free - Drawing Fun     ART_AND_DESIGN   
10                            Text on Photo - Fonteee     ART_AND_DESIGN   
11            Name Art Photo Editor - Focus n Filters     ART_AND_DESIGN   
12                     Tattoo Name On My Photo Editor     ART_AND_DESIGN   
13                              Mandala Coloring Book     ART_AND_DESIGN   
14    3D Color Pixel by Number - Sandbox Art Coloring     ART_AND_DESIGN   
15                    Learn To Draw Kawaii Characters     ART_AND_DESIGN   
16       Photo Designer - Write your name with shapes     ART_AND_DESIGN   
17                           350 Diy Room Decor Ideas     ART_AND_DESIGN   
18                      FlipaClip - Cartoon animation     ART_AND_DESIGN   
19                                       ibis Paint X     ART_AND_DESIGN   
20                        Logo Maker - Small Business     ART_AND_DESIGN   
21          Boys Photo Editor - Six Pack & Men's Suit     ART_AND_DESIGN   
22            Superheroes Wallpapers | 4K Backgrounds     ART_AND_DESIGN   
23                             Mcqueen Coloring pages     ART_AND_DESIGN   
24                        HD Mickey Minnie Wallpapers     ART_AND_DESIGN   
25                         Harley Quinn wallpapers HD     ART_AND_DESIGN   
26                      Colorfit - Drawing & Coloring     ART_AND_DESIGN   
27                              Animated Photo Editor     ART_AND_DESIGN   
28                              Pencil Sketch Drawing     ART_AND_DESIGN   
29                    Easy Realistic Drawing Tutorial     ART_AND_DESIGN   
..                                                ...                ...   
70  Fines of the State Traffic Safety Inspectorate...  AUTO_AND_VEHICLES   
71            SK Enca Direct Malls - Used Cars Search  AUTO_AND_VEHICLES   
72      Android Auto - Maps, Media, Messaging & Voice  AUTO_AND_VEHICLES   
73                                             PDD-UA  AUTO_AND_VEHICLES   
74                         Tickets SDA 2019 + Exam RF  AUTO_AND_VEHICLES   
75              Super Cars Wallpapers And Backgrounds  AUTO_AND_VEHICLES   
76                  Police Lights, Sirens & Follow Me  AUTO_AND_VEHICLES   
77               Police Detector (Speed Camera Radar)  AUTO_AND_VEHICLES   
78                                Best Car Wallpapers  AUTO_AND_VEHICLES   
79                            Tickets + PDA 2018 Exam  AUTO_AND_VEHICLES   
80                              Pick Your Part Garage  AUTO_AND_VEHICLES   
81                         PakWheels: Buy & Sell Cars  AUTO_AND_VEHICLES   
82                                Supervision service  AUTO_AND_VEHICLES   
83      Speed Camera Detector - Traffic & Speed Alert  AUTO_AND_VEHICLES   
84        Used car search Goo net whole car Go to net  AUTO_AND_VEHICLES   
85  CarMax – Cars for Sale: Search Used Car Inventory  AUTO_AND_VEHICLES   
86                                    BEST CAR SOUNDS  AUTO_AND_VEHICLES   
87                      RST - Sale of cars on the PCT  AUTO_AND_VEHICLES   
88        AutoScout24 Switzerland – Find your new car  AUTO_AND_VEHICLES   
89  Zona Azul Digital Fácil SP CET - OFFICIAL São ...  AUTO_AND_VEHICLES   
90                                           SMS Park  AUTO_AND_VEHICLES   
91                    

In [105]:
gps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
App               10841 non-null object
Category          10841 non-null object
Rating            9367 non-null float64
Reviews           10841 non-null object
Size              10841 non-null object
Installs          10841 non-null object
Type              10840 non-null object
Price             10841 non-null object
Content Rating    10840 non-null object
Genres            10841 non-null object
Last Updated      10841 non-null object
Current Ver       10833 non-null object
Android Ver       10838 non-null object
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


** NO. 1 : there are 13 columns and 10841 row 

** DATA CLEANSING - Remove duplicates App

In [106]:
gdrop = gps.drop_duplicates(['App'], keep='first')

In [107]:
gdrop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9660 entries, 0 to 10840
Data columns (total 13 columns):
App               9660 non-null object
Category          9660 non-null object
Rating            8197 non-null float64
Reviews           9660 non-null object
Size              9660 non-null object
Installs          9660 non-null object
Type              9659 non-null object
Price             9660 non-null object
Content Rating    9659 non-null object
Genres            9660 non-null object
Last Updated      9660 non-null object
Current Ver       9652 non-null object
Android Ver       9657 non-null object
dtypes: float64(1), object(12)
memory usage: 1.0+ MB


** NO. 2 : there are 34 unique App

In [108]:
len(gdrop["Category"].unique())

34

In [109]:
len(gdrop["Genres"].unique())

119

** DATA CLEANSING - Remove unwanted characters ('+' and 'FREE') in Installs Columns

In [120]:
gdrop['Installs'] = gdrop['Installs'].map(lambda x: x.lstrip('+-').rstrip('FfRrEe'))

AttributeError: 'float' object has no attribute 'lstrip'

In [111]:
gdrop['Installs'] = gdrop['Installs'].str.replace(r'\D', '')

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [121]:
gdrop.head (5)

App             Category  Rating   Reviews  \
336          WhatsApp Messenger        COMMUNICATION     4.4  69119316   
152           Google Play Books  BOOKS_AND_REFERENCE     3.9   1433233   
1654             Subway Surfers                 GAME     4.5  27722264   
3117  Maps - Navigate & Explore     TRAVEL_AND_LOCAL     4.3   9235155   
340                       Gmail        COMMUNICATION     4.3   4604324   

                    Size      Installs  Type Price Content Rating  \
336   Varies with device  1.000000e+09  Free     0       Everyone   
152   Varies with device  1.000000e+09  Free     0           Teen   
1654                 76M  1.000000e+09  Free     0   Everyone 10+   
3117  Varies with device  1.000000e+09  Free     0       Everyone   
340   Varies with device  1.000000e+09  Free     0       Everyone   

                 Genres    Last Updated         Current Ver  \
336       Communication  August 3, 2018  Varies with device   
152   Books & Reference  August 3, 2018  Varies with device   
1654             Arcade   July 12, 2018              1.90.0   
3117     Travel & Local   July 31, 2018  Varies with device   
340       Communication  August 2, 2018  Varies with device   

             Android Ver  
336   Varies with device  
152   Varies with device  
1654          4.1 and up  
3117  Varies with device  
340   Varies with device

** DATA MANIPULATION - Change 'Intalls' column type from object to numeric

In [127]:
gdrop["Installs"] = pd.to_numeric(gdrop["Installs"])

In [128]:
gdrop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9660 entries, 336 to 10472
Data columns (total 13 columns):
App               9660 non-null object
Category          9660 non-null object
Rating            8197 non-null float64
Reviews           9660 non-null object
Size              9660 non-null object
Installs          9659 non-null float64
Type              9659 non-null object
Price             9660 non-null object
Content Rating    9659 non-null object
Genres            9660 non-null object
Last Updated      9660 non-null object
Current Ver       9652 non-null object
Android Ver       9657 non-null object
dtypes: float64(2), object(11)
memory usage: 1.0+ MB


** Sorting table by Installs App and Group it By Category column

In [129]:
gdrop = gdrop.sort_values(by ='Installs' , ascending = False)

In [130]:
cat = gdrop.groupby('Category')

In [131]:
gdrop.head (3)

App          Category  Rating   Reviews  \
336   WhatsApp Messenger     COMMUNICATION     4.4  69119316   
3127  Google Street View  TRAVEL_AND_LOCAL     4.2   2129689   
3234              Google             TOOLS     4.4   8033493   

                    Size      Installs  Type Price Content Rating  \
336   Varies with device  1.000000e+09  Free     0       Everyone   
3127  Varies with device  1.000000e+09  Free     0       Everyone   
3234  Varies with device  1.000000e+09  Free     0       Everyone   

              Genres    Last Updated         Current Ver         Android Ver  
336    Communication  August 3, 2018  Varies with device  Varies with device  
3127  Travel & Local  August 6, 2018  Varies with device  Varies with device  
3234           Tools  August 3, 2018  Varies with device  Varies with device

** Whats App Messenger, Google Play Books, and Subway Surfers are top 3 App based on Installs

In [132]:
for Category,Category_gdrop in cat :
    print(Category)
    print()
    print(Category_gdrop[['App','Installs','Rating']].sort_values(by ='Installs' , ascending = False).head(20))
    print()

1.9

                                           App  Installs  Rating
10472  Life Made WI-Fi Touchscreen Photo Frame       NaN    19.0

ART_AND_DESIGN

                                                  App    Installs  Rating
3                               Sketch - Draw & Paint  50000000.0     4.5
45  Canva: Poster, banner, card maker & graphic de...  10000000.0     4.7
12                     Tattoo Name On My Photo Editor  10000000.0     4.2
19                                       ibis Paint X  10000000.0     4.6
42                         Textgram - write on photos  10000000.0     4.4
18                      FlipaClip - Cartoon animation   5000000.0     4.3
37                                 Floor Plan Creator   5000000.0     4.1
2   U Launcher Lite – FREE Live Cool Themes, Hide ...   5000000.0     4.7
10                            Text on Photo - Fonteee   1000000.0     4.4
8                                Garden Coloring Book   1000000.0     4.4
11            Name Art Photo Edito

3855                                         Blitzer.de    5000000.0     4.5

MEDICAL

                                                    App   Installs  Rating
2310                                     Blood Pressure  5000000.0     4.2
2319                       My Calendar - Period Tracker  5000000.0     4.7
2322                         1800 Contacts - Lens Store  1000000.0     4.7
2510                           Blood Pressure(BP) Diary  1000000.0     3.7
2307   Ovia Pregnancy Tracker & Baby Countdown Calendar  1000000.0     4.8
2316      Ovia Fertility Tracker & Ovulation Calculator  1000000.0     4.8
2312                                   Doctor On Demand  1000000.0     4.7
2311                             Pregnancy Week By Week  1000000.0     4.8
2333               Pregnancy Calculator and Tracker app  1000000.0     4.8
2305                                    FollowMyHealth®  1000000.0     4.6
2306                                           CareZone  1000000.0     4.4
2304         

In [119]:
gdrop_r = gdrop.sort_values(by ='Rating' , ascending = False)

In [30]:
gdrop_r.head (5)

App   Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame        1.9    19.0    3.0M   
8230                                   DB HOME  LIFESTYLE     5.0       5   
9518                   Asha Ek Hope - ALS/ MND    MEDICAL     5.0       2   
7872                         CT Cervical Spine    MEDICAL     5.0       5   
7881                      CARDIAC CT TECHNIQUE    MEDICAL     5.0       6   

         Size  Installs  Type     Price Content Rating             Genres  \
10472  1,000+       NaN     0  Everyone            NaN  February 11, 2018   
8230     3.2M     100.0  Free         0       Everyone          Lifestyle   
9518      11M     100.0  Free         0       Everyone            Medical   
7872      17M    1000.0  Free         0       Everyone            Medical   
7881      17M    1000.0  Free         0       Everyone            Medical   

           Last Updated Current Ver   Android Ver  
10472            1.0.19  4.0 and up           NaN  
8230    August 31, 2017       1.0.1    4.0 and up  
9518       May 23, 2018       1.1.0    4.1 and up  
7872   January 20, 2018         2.0    4.0 and up  
7881     March 10, 2018         2.0  4.0.3 and up

In [31]:
rat = gdrop_r.groupby('Category')

In [32]:
for Category,Category_gdrop_r in rat :
    print(Category)
    print()
    print(Category_gdrop_r[['App','Rating']].sort_values(by ='Rating' , ascending = False).head(3))
    print()

1.9

                                           App  Rating
10472  Life Made WI-Fi Touchscreen Photo Frame    19.0

ART_AND_DESIGN

                                          App  Rating
8888  Spring flowers theme couleurs d t space     5.0
6930     Fantasy theme dark bw black building     4.8
25                 Harley Quinn wallpapers HD     4.8

AUTO_AND_VEHICLES

                                                    App  Rating
4120                                       MHD F-Series     4.9
64                DMV Permit Practice Test 2018 Edition     4.9
55    Tickets SDA 2018 and Exam from the State Traff...     4.9

BEAUTY

                                 App  Rating
99    ipsy: Makeup, Beauty, and Tips     4.9
108             Prom MakeUp Tutorial     4.8
100  Natural recipes for your beauty     4.7

BOOKS_AND_REFERENCE

                             App  Rating
5763  Tozer Devotional -Series 1     5.0
5237      Hey AJ! It's Saturday!     5.0
6840       Catholic La Bu Zo Kam     5.0



In [33]:
gdrop[["Installs"]].describe()


Installs
count  9.659000e+03
mean   7.777507e+06
std    5.375828e+07
min    0.000000e+00
25%    1.000000e+03
50%    1.000000e+05
75%    1.000000e+06
max    1.000000e+09

In [34]:
gdrop[["Rating"]].describe()

Rating
count  8197.000000
mean      4.175052
std       0.561026
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [36]:
gcomb = gdrop[(gdrop.Rating > 4.5) & (gdrop.Installs > 1000000) ]

In [37]:
cat = gcomb.groupby('Category')

In [74]:
for Category,Category_gcomb in cat :
    print(Category)
    print()
    print(Category_gcomb[['App','Installs','Rating']].sort_values(by ='Installs' , ascending = False).head(20))
    print()

ART_AND_DESIGN

                                                  App    Installs  Rating
19                                       ibis Paint X  10000000.0     4.6
45  Canva: Poster, banner, card maker & graphic de...  10000000.0     4.7
2   U Launcher Lite – FREE Live Cool Themes, Hide ...   5000000.0     4.7

AUTO_AND_VEHICLES

                                                  App   Installs  Rating
70  Fines of the State Traffic Safety Inspectorate...  5000000.0     4.8

BOOKS_AND_REFERENCE

                                                    App     Installs  Rating
139                                Wattpad 📖 Free Books  100000000.0     4.6
3941                                              Bible  100000000.0     4.7
6290  Dictionary.com: Find Definitions for English W...   10000000.0     4.6
5345                                  Quran for Android   10000000.0     4.7
9625                                         JW Library   10000000.0     4.9
5319                                  

In [77]:
gdrop["Category"].unique()

array(['COMMUNICATION', 'BOOKS_AND_REFERENCE', 'GAME', 'TRAVEL_AND_LOCAL',
       'SOCIAL', 'VIDEO_PLAYERS', 'TOOLS', 'PRODUCTIVITY',
       'ENTERTAINMENT', 'PHOTOGRAPHY', 'NEWS_AND_MAGAZINES',
       'HEALTH_AND_FITNESS', 'FAMILY', 'LIFESTYLE', 'SHOPPING',
       'PERSONALIZATION', 'BUSINESS', 'FINANCE', 'SPORTS', 'EDUCATION',
       'MAPS_AND_NAVIGATION', 'ART_AND_DESIGN', 'WEATHER', 'PARENTING',
       'AUTO_AND_VEHICLES', 'LIBRARIES_AND_DEMO', 'HOUSE_AND_HOME',
       'FOOD_AND_DRINK', 'COMICS', 'DATING', 'BEAUTY', 'MEDICAL',
       'EVENTS', '1.9'], dtype=object)

** DATA MANIPULATION - Change 'Reviews' column type from object to numeric, replace 'K' 'M' with int)

In [142]:
gdrop.Reviews = (gdrop.Reviews.replace(r'[KM]+$', '', regex=True).astype(float) * \
                 gdrop.Reviews.str.extract(r'[\d\.]+([KM]+)', expand=False)
                 .fillna(1)
                 .replace(['K','M'], [10**3, 10**6]).astype(int))

** Sorting table by Reviews and Group it By Category column

In [143]:
gdrop_reviews = gdrop.sort_values(by ='Reviews' , ascending = False)

In [144]:
gdrop_reviews.head()

App       Category  Rating  \
2544                                  Facebook         SOCIAL     4.1   
336                         WhatsApp Messenger  COMMUNICATION     4.4   
2545                                 Instagram         SOCIAL     4.5   
335   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
1670                            Clash of Clans           GAME     4.6   

         Reviews                Size      Installs  Type Price Content Rating  \
2544  78158306.0  Varies with device  1.000000e+09  Free     0           Teen   
336   69119316.0  Varies with device  1.000000e+09  Free     0       Everyone   
2545  66577313.0  Varies with device  1.000000e+09  Free     0           Teen   
335   56642847.0  Varies with device  1.000000e+09  Free     0       Everyone   
1670  44891723.0                 98M  1.000000e+08  Free     0   Everyone 10+   

             Genres    Last Updated         Current Ver         Android Ver  
2544         Social  August 3, 2018  Varies with device  Varies with device  
336   Communication  August 3, 2018  Varies with device  Varies with device  
2545         Social   July 31, 2018  Varies with device  Varies with device  
335   Communication  August 1, 2018  Varies with device  Varies with device  
1670       Strategy   July 15, 2018           10.322.16          4.1 and up

In [145]:
rev = gdrop_reviews.groupby('Category')

In [146]:
for Category,Category_gdrop_reviews in rev  :
    print(Category)
    print()
    print(Category_gdrop_reviews[['App','Rating','Reviews']].sort_values(by ='Reviews' , ascending = False).head(20))
    print()

1.9

                                           App  Rating    Reviews
10472  Life Made WI-Fi Touchscreen Photo Frame    19.0  3000000.0

ART_AND_DESIGN

                                                    App  Rating   Reviews
42                           Textgram - write on photos     4.4  295221.0
19                                         ibis Paint X     4.6  224399.0
3                                 Sketch - Draw & Paint     4.5  215644.0
18                        FlipaClip - Cartoon animation     4.3  194216.0
45    Canva: Poster, banner, card maker & graphic de...     4.7  174531.0
2     U Launcher Lite – FREE Live Cool Themes, Hide ...     4.7   87510.0
12                       Tattoo Name On My Photo Editor     4.2   44829.0
7                                      Infinite Painter     4.1   36815.0
37                                   Floor Plan Creator     4.1   36639.0
26                        Colorfit - Drawing & Coloring     4.7   20260.0
10                              

                                                    App  Rating    Reviews
2655                           Wish - Shopping Made Fun     4.5  6210998.0
2698                       Flipkart Online Shopping App     4.4  6012719.0
2660       AliExpress - Smarter Shopping, Better Living     4.6  5916606.0
2684           Mercado Libre: Find your favorite brands     4.7  3860225.0
2662  eBay: Buy & Sell this Summer - Discover Deals ...     4.4  2788923.0
2665                   Lazada - Online Shopping & Deals     4.2  1573054.0
2716          Groupon - Shop Deals, Discounts & Coupons     4.6  1370749.0
2693                         Myntra Online Shopping App     4.3  1315242.0
2674                                          The birth     4.7  1084945.0
2663   letgo: Buy & Sell Used Stuff, Cars & Real Estate     4.5   973270.0
2664                                    Amazon Shopping     4.3   909226.0
2666                                 OLX - Buy and Sell     4.2   857923.0
2679                    B

In [147]:
gdrop_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9660 entries, 2544 to 10692
Data columns (total 13 columns):
App               9660 non-null object
Category          9660 non-null object
Rating            8197 non-null float64
Reviews           9660 non-null float64
Size              9660 non-null object
Installs          9659 non-null float64
Type              9659 non-null object
Price             9660 non-null object
Content Rating    9659 non-null object
Genres            9660 non-null object
Last Updated      9660 non-null object
Current Ver       9652 non-null object
Android Ver       9657 non-null object
dtypes: float64(3), object(10)
memory usage: 1.0+ MB
